pallets/parachain-staking/src/migrations.rs

```diff
@@ -15,12 +15,169 @@
 // along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
 
 //! # Migrations
-use crate::{Config, Points, Round, Staked};
+use crate::{
+	pallet::{migrate_nominator_to_delegator_state, RoundIndex},
+	BalanceOf, CandidateState, CollatorCandidate, CollatorState2, Config, DelegatorState,
+	ExitQueue2, NominatorState2, Points, Round, Staked,
+};
 use frame_support::{
 	pallet_prelude::PhantomData,
 	traits::{Get, OnRuntimeUpgrade},
 	weights::Weight,
 };
+use sp_std::collections::btree_map::BTreeMap;
+
+/// Migration to replace the automatic ExitQueue with a manual exits API.
+/// This migration is idempotent so it can be run more than once without any risk.
+/// Returns (reads, writes)
+pub struct RemoveExitQueue<T>(PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for RemoveExitQueue<T> {
+	fn on_runtime_upgrade() -> Weight {
+		log::info!(target: "RemoveExitQueue", "running migration to remove staking exit queue");
+		let exit_queue = <ExitQueue2<T>>::take();
+		let (mut reads, mut writes) = (1u64, 0u64);
+		let mut delegator_exits: BTreeMap<T::AccountId, RoundIndex> = BTreeMap::new();
+		let mut delegation_revocations: BTreeMap<T::AccountId, (T::AccountId, RoundIndex)> =
+			BTreeMap::new();
+		// Track scheduled delegator exits and revocations before migrating state
+		// Candidates already track exit info locally so no tracking is necessary
+		for (delegator, is_revocation, when) in exit_queue.nominator_schedule {
+			if let Some(revoking_candidate) = is_revocation {
+				delegation_revocations.insert(delegator, (revoking_candidate, when));
+			} else {
+				delegator_exits.insert(delegator, when);
+			}
+		}
+		// execute candidate migration
+		for (candidate_id, collator_state) in <CollatorState2<T>>::drain() {
+			let candidate_state: CollatorCandidate<T::AccountId, BalanceOf<T>> =
+				collator_state.into();
+			<CandidateState<T>>::insert(candidate_id, candidate_state);
+			reads += 1u64;
+			writes += 1u64;
+		}
+		// execute delegator migration
+		for (delegator_id, nominator_state) in <NominatorState2<T>>::drain() {
+			let mut delegator_state =
+				migrate_nominator_to_delegator_state::<T>(delegator_id.clone(), nominator_state);
```

Comment:
This could be pretty heavy CPU-wise, I wonder if your 10% estimate might be insufficient here...

Reply:
This method is a custom From conversion from struct Nominator2 to struct Delegator which also takes 1 parameter. So I don't think the conversion itself is expensive. 

I'm more worried about iterating all the items in the NominatorState2 map, maybe that's what you were referring to?

Either way, we can increase the margin of safety.

Reply:
increased to 50%

Reply:
Yeah, it looks like we can expect around 3k of them. I think we'll be ok, but (as I mentioned privately) I'll help you test this with `try-runtime`.